src/core/search/compressed_sorted_set.h

```diff
@@ -106,7 +81,8 @@ class CompressedSortedSet {
   static std::pair<IntType /*value*/, size_t /*read*/> ReadVarLen(absl::Span<const uint8_t> source);
 
  private:
-  IntType size_{0};
+  uint32_t size_{0};
+  std::optional<IntType> tail_value_{};
```

Comment:
Currently, it seems like you only update this var on write (insert, remove). You could also update it inside `ReadNext()` for the cases where it is empty and you stumble across the last entry.
This whole thing is an optimization, so I'm not at all sure if it's worth it. I'll leave it to you do consider complexity vs frequency of it happening :)

Reply:
That's certainly a good idea - though it'd require growing the iterator to have a pointer to the sorted set so it can update it. I'll follow up with this in a separate optimizations PR